plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -127,3 +154,235 @@ counted out already so we should not immediately encounter those in `VBody`.
 Also, we currently reject `Constant` (has acceptable cost but not acceptable size).
 We may want to check their sizes instead of just rejecting them.
 -}
+
+-- | Inline fully applied functions. Note [Inlining of fully applied functions].
+callSiteInline
+    :: forall tyname name uni fun a m
+    . (ExternalConstraints tyname name uni fun m, Eq a, Ord name)
+    => InlineHints name a
+    -> PLC.BuiltinVersion fun
+    -> Term tyname name uni fun a
+    -> m (Term tyname name uni fun a)
+callSiteInline hints ver t = let
+        inlineInfo :: InlineInfo name fun a
+        inlineInfo = InlineInfo (snd deps) usgs hints ver
+        -- We actually just want the variable strictness information here!
+        deps :: (G.Graph Deps.Node, Map.Map PLC.Unique Strictness)
+        deps = Deps.runTermDeps ver t
+        usgs :: Usages.Usages
+        usgs = Usages.termUsages t
+    in liftQuote $ flip evalStateT mempty $ flip runReaderT inlineInfo $ processTerm t
+
+-- | Run the inliner on a `Core.Type.Term`.
+processTerm
+    :: forall tyname name uni fun a. (InliningConstraints tyname name uni fun, Eq a, Ord name)
+    => Term tyname name uni fun a -- ^ Term to be processed.
+    -> InlineM tyname name uni fun a (Term tyname name uni fun a)
+processTerm = \case
+    -- When we encounter a variable, consider inlining it.
+    v@(Var _a _n) -> do
+      considerInline v
+    Let a NonRec bs t -> do
+        -- Process non-recursive let bindings, checking for fully applied functions and
+        -- put them in the substitution. We do NOT remove the inlined binding here because it
+        -- may not be dead. We leave this to the dead code pass.
+        bs' <- traverse (processSingleBinding t) bs
```

Comment:
Right! Because inlining may reduce size.

Reply:
Although it won't happen here right, it will only happen after beta reduction? Anyway this is easy to fix so we'll do that.

Reply:
I don't see why we would process them in reverse? I think the code that's already there is right? Do you have an example?

Reply:
I just realized you are talking about the *binding* order. When I made the replies I was thinking about inner variables: the inner one should be inlined first, then, if its size can be reduced, it may enable the outer one to be inlined. But only after beta reduction. So yeah I agree with Michael.

Reply:
I mean, in `let x = ..., y = ..x.. in body`, if you process `y` first, could it possibly make it easier to inline `x`? I don't know that it can, but I also don't know that it definitely cannot.

The reverse order cannot possibly be wrong, can it?

Reply:
E.g., say `x` is not fully applied in `y`'s RHS. But after you inline `y`, now `x` becomes fully saturated, and you can inline `x`!

Reply:
I still don't see how from you eg but like you said, it can't hurt, so I can do that.

Reply:
This seems the wrong way around. Typically, it's easier to optimize terms that more things have been inlined _into_. That argues for processing things in dependency order, which is what we're already doing.

We _definitely_ shouldn't introduce quirks of behaviour like this without clear examples of where it helps!